File: 63. Unique Paths II.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        vector<int> ans(obstacleGrid[0].size(), 0);
        ans[0] = 1 - obstacleGrid[0][0];
        for(int i = 0; i < obstacleGrid.size(); ++i) {
            for(int j = 0; j < obstacleGrid[0].size(); ++j) {
                if(obstacleGrid[i][j] == 1) ans[j] = 0;
                else ans[j] += (j >= 1)? ans[j - 1] : 0;
            }
        }
        return ans.back();
        
    }
};

File: 63. Unique Paths II.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """
        n, m = len(obstacleGrid) + 1, len(obstacleGrid[0]) + 1
        ans = [[0] * m for _ in range(n)] 
        ans[1][1] = 1 - obstacleGrid[0][0]
        for i in range(1, n):
            for j in range(1, m):
                if obstacleGrid[i - 1][j - 1] or (i == 1 and j == 1):
                    continue
                ans[i][j] = ans[i - 1][j] + ans[i][j - 1]
        return ans[-1][-1]
        